File: ios/Runner/SpeechTranscriptionPlugin.swift
import Flutter
import Speech

public class SpeechTranscriptionPlugin: NSObject, FlutterPlugin {
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "com.seedling.speech_transcription",
            binaryMessenger: registrar.messenger()
        )
        let instance = SpeechTranscriptionPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "isAvailable":
            checkAvailability(result: result)
        case "transcribe":
            guard let args = call.arguments as? [String: Any],
                  let filePath = args["filePath"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "filePath required", details: nil))
                return
            }
            let locale = args["locale"] as? String
            transcribeFile(filePath: filePath, locale: locale, result: result)
        case "requestPermission":
            requestPermission(result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func checkAvailability(result: @escaping FlutterResult) {
        let recognizer = SFSpeechRecognizer()
        let isAvailable = recognizer?.isAvailable ?? false
        let supportsOnDevice: Bool
        if #available(iOS 13.0, *) {
            supportsOnDevice = recognizer?.supportsOnDeviceRecognition ?? false
        } else {
            supportsOnDevice = false
        }
        result([
            "isAvailable": isAvailable,
            "supportsOnDevice": supportsOnDevice,
            "authStatus": SFSpeechRecognizer.authorizationStatus().rawValue
        ])
    }

    private func requestPermission(result: @escaping FlutterResult) {
        SFSpeechRecognizer.requestAuthorization { status in
            DispatchQueue.main.async {
                result([
                    "authorized": status == .authorized,
                    "status": status.rawValue
                ])
            }
        }
    }

    private func transcribeFile(filePath: String, locale: String?, result: @escaping FlutterResult) {
        let recognizerLocale = locale != nil ? Locale(identifier: locale!) : nil
        let recognizer: SFSpeechRecognizer?
        if let loc = recognizerLocale {
            recognizer = SFSpeechRecognizer(locale: loc)
        } else {
            recognizer = SFSpeechRecognizer()
        }

        guard let speechRecognizer = recognizer, speechRecognizer.isAvailable else {
            result(FlutterError(code: "UNAVAILABLE", message: "Speech recognition unavailable", details: nil))
            return
        }

        let url = URL(fileURLWithPath: filePath)
        let request = SFSpeechURLRecognitionRequest(url: url)

        // Prefer on-device recognition for privacy
        if #available(iOS 13.0, *) {
            request.requiresOnDeviceRecognition = speechRecognizer.supportsOnDeviceRecognition
        }

        speechRecognizer.recognitionTask(with: request) { [weak self] taskResult, error in
            guard self != nil else { return }

            if let error = error {
                DispatchQueue.main.async {
                    result(FlutterError(
                        code: "TRANSCRIPTION_ERROR",
                        message: error.localizedDescription,
                        details: nil
                    ))
                }
                return
            }

            guard let taskResult = taskResult else { return }

            if taskResult.isFinal {
                let transcription = taskResult.bestTranscription.formattedString
                let segments = taskResult.bestTranscription.segments.map { segment -> [String: Any] in
                    return [
                        "substring": segment.substring,
                        "timestamp": segment.timestamp,
                        "duration": segment.duration,
                        "confidence": segment.confidence
                    ]
                }
                DispatchQueue.main.async {
                    result([
                        "transcription": transcription,
                        "segments": segments,
                        "isFinal": true
                    ])
                }
            }
        }
    }
}

File: android/app/src/main/kotlin/com/seedling/seedling/SpeechTranscriptionPlugin.kt
package com.twotwoeightthreelabs.seedling

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.speech.RecognitionListener
import android.speech.RecognizerIntent
import android.speech.SpeechRecognizer
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import android.media.MediaPlayer
import java.io.File

class SpeechTranscriptionPlugin : FlutterPlugin, MethodCallHandler {
    private lateinit var channel: MethodChannel
    private lateinit var context: Context

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(binding.binaryMessenger, "com.seedling.speech_transcription")
        channel.setMethodCallHandler(this)
        context = binding.applicationContext
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "isAvailable" -> checkAvailability(result)
            "transcribe" -> {
                val filePath = call.argument<String>("filePath")
                if (filePath == null) {
                    result.error("INVALID_ARGS", "filePath required", null)
                    return
                }
                val locale = call.argument<String>("locale")
                transcribeFile(filePath, locale, result)
            }
            "requestPermission" -> {
                // Android speech recognition doesn't need separate permission
                // (uses RECORD_AUDIO which is already granted for voice recording)
                result.success(mapOf("authorized" to true, "status" to 0))
            }
            else -> result.notImplemented()
        }
    }

    private fun checkAvailability(result: Result) {
        val isAvailable = SpeechRecognizer.isRecognitionAvailable(context)
        val supportsOnDevice = if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.S) {
            SpeechRecognizer.isOnDeviceRecognitionAvailable(context)
        } else {
            false
        }
        result.success(mapOf(
            "isAvailable" to isAvailable,
            "supportsOnDevice" to supportsOnDevice,
            "authStatus" to if (isAvailable) 3 else 0
        ))
    }

    private fun transcribeFile(filePath: String, locale: String?, result: Result) {
        val file = File(filePath)
        if (!file.exists()) {
            result.error("FILE_NOT_FOUND", "Audio file not found: $filePath", null)
            return
        }

        if (!SpeechRecognizer.isRecognitionAvailable(context)) {
            result.error("UNAVAILABLE", "Speech recognition unavailable", null)
            return
        }

        val recognizer = SpeechRecognizer.createSpeechRecognizer(context)
        val intent = Intent(RecognizerIntent.ACTION_RECOGNIZE_SPEECH).apply {
            putExtra(RecognizerIntent.EXTRA_LANGUAGE_MODEL, RecognizerIntent.LANGUAGE_MODEL_FREE_FORM)
            if (locale != null) {
                putExtra(RecognizerIntent.EXTRA_LANGUAGE, locale)
            }
            putExtra(RecognizerIntent.EXTRA_PARTIAL_RESULTS, false)
            // Note: Android SpeechRecognizer works with microphone input.
            // For file-based transcription, we use MediaPlayer workaround
            // or fall back to reporting unavailability for file transcription.
        }

        recognizer.setRecognitionListener(object : RecognitionListener {
            override fun onResults(results: Bundle?) {
                val matches = results?.getStringArrayList(SpeechRecognizer.RESULTS_RECOGNITION)
                val confidences = results?.getFloatArray(SpeechRecognizer.CONFIDENCE_SCORES)
                val transcription = matches?.firstOrNull() ?: ""

                result.success(mapOf(
                    "transcription" to transcription,
                    "segments" to emptyList<Map<String, Any>>(),
                    "isFinal" to true
                ))
                recognizer.destroy()
            }

            override fun onError(error: Int) {
                val message = when (error) {
                    SpeechRecognizer.ERROR_AUDIO -> "Audio recording error"
                    SpeechRecognizer.ERROR_CLIENT -> "Client side error"
                    SpeechRecognizer.ERROR_INSUFFICIENT_PERMISSIONS -> "Insufficient permissions"
                    SpeechRecognizer.ERROR_NETWORK -> "Network error"
                    SpeechRecognizer.ERROR_NETWORK_TIMEOUT -> "Network timeout"
                    SpeechRecognizer.ERROR_NO_MATCH -> "No speech detected"
                    SpeechRecognizer.ERROR_RECOGNIZER_BUSY -> "Recognizer busy"
                    SpeechRecognizer.ERROR_SERVER -> "Server error"
                    SpeechRecognizer.ERROR_SPEECH_TIMEOUT -> "No speech input"
                    else -> "Unknown error: $error"
                }
                result.error("TRANSCRIPTION_ERROR", message, null)
                recognizer.destroy()
            }

            override fun onReadyForSpeech(params: Bundle?) {}
            override fun onBeginningOfSpeech() {}
            override fun onRmsChanged(rmsdB: Float) {}
            override fun onBufferReceived(buffer: ByteArray?) {}
            override fun onEndOfSpeech() {}
            override fun onPartialResults(partialResults: Bundle?) {}
            override fun onEvent(eventType: Int, params: Bundle?) {}
        })

        recognizer.startListening(intent)
    }
}
